apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -98,6 +101,7 @@ open_session(ClientID) ->
 add_subscription(TopicFilterBin, DSSessionID) ->
     ?WHEN_ENABLED(
         begin
+            ok = emqx_persistent_session_ds_router:do_add_route(TopicFilterBin, DSSessionID),
```

Comment:
Q: This executes out of transaction context, right? If so, it seems there's possibility for an inconsistency: route is present in the persistent session router, but not in the session state. Admittedly, I might be missing something here.

Same question about `del_subscripton/2`.

Reply:
Indeed, subscriptions are (serially) processed by the connection process without any locking/transaction context.

I think I see what you mean.  Do you mean that the inconsistency might arise from the route is inserted into the PS routing table, and then the node crashes?  

Good point, perhaps it would be safer to first ensure the iterator is added to the session before adding the route to make it idempotent when restarting the subscription.

Reply:
> perhaps it would be safer to first ensure the iterator is added to the session before adding the route to make it idempotent when restarting the subscription.

Indeed, it should be better. But there's also a possibility that no client will ever try to restart the subscription again.

Reply:
Yes, there's that possibility.

By writing to the iterator reference table first, inside a transaction, I was thinking that it could cover that scenario when the client connects again.  My idea was that the client, upon reconnecting, would have its session data loaded, and when encountering that topic associated with its DS session ID, it would also call `add_subscription` to repeat the process. 

Also, we're currently using dirty operations for writing the routes, so there's no use in calling it inside `emqx_ds:session_add_iterator`'s transaction, for example.

Do you see a way to make this more consistent?  :thinking:

Reply:
>it would also call add_subscription to repeat the process.

Actually, I was thinking of `emqx_session:subscribe`.  On second thought, I wonder if this is redundant/overkill...  :thinking:

Reply:
Good points.

Under the assumption that the client will eventually attempt to repeat any failed operation until successful completion, we indeed should get away with simple sequence of writes that's only idempotent (but not transactional). But it's very strong assumption that probably won't hold in practice.

> Do you see a way to make this more consistent? 🤔

Only 2 things come to mind so far.
1. Wrap all subscription-related writes in a transaction.
2. Turn session router into an asynchronous in-memory index of the sessions subscriptions table.

On the other hand, the risk of having inconsistent state between sessions and session router tables doesn't compromise correctness (not 100% sure, please tell if you agree): i.e. we only risk persisting messages that no client will ever read. So, I think it might be fine for now, but we'll probably need to think up a way how to reconcile this inconsistency eventually.

> On second thought, I wonder if this is redundant/overkill...

In general, for a transactionless idempotent multi-record write we supposedly should designate a kind of a "write root", a record that's written the last, and the existence of which can later tell that the whole write sequence was successful, so we wouldn't need to repeat (and a way to tell if a record is part of incomplete write, in order to GC those records eventually).

With subscriptions, "write root" is probably session's subscription record in mnesia. So write sequence supposedly should be _iterators_ -> _session route_ -> _session subscription record_. Thus we will know for sure that, if this _session subscription record_ exists, everything written before it also exists.

These are general thoughts, and this scheme is likely not universally applicable. Yet, should be fine as a kind of coarse mental model. What do you think?

Reply:
> perhaps it would be safer to first ensure the iterator is added to the session before adding the route to make it idempotent when restarting the subscription.

Sorry for the confusion. AFAICT it's more correct the way it currently is in this PR.

I.e. without transactions (or some other mechanism) we cannot hope for consistency anyway, so we should write the route record before subscription record. Otherwise, the client will not see messages written after `StartMS` but before it reestablishes the connection (and optionally resubscribes).